prometheus_hardware_exporter/collector.py

```diff
@@ -5,12 +5,134 @@
+                documentation="Total number of controllers",
+                labels=["hostname"],
+                metric_class=GaugeMetricFamily,
+            ),
+            Specification(
+                name="poweredgeraid_virtual_drive",
+                documentation="Number of virtual drives",
+                labels=["controller_id"],
+                metric_class=GaugeMetricFamily,
+            ),
+            Specification(
+                name="perccli_command_success",
+                documentation="Indicates the if command is successful or not",
+                labels=[],
+                metric_class=GaugeMetricFamily,
+            ),
+            Specification(
+                name="perccli_command_ctrl_success",
+                documentation="Indicates the if command for each controller is successful or not",
+                labels=["controller_id"],
+                metric_class=GaugeMetricFamily,
+            ),
+            Specification(
+                name="poweredgeraid_virtual_drive_state",
+                documentation="Indicates the state of virtual drive",
+                labels=["controller_id", "device_group", "virtual_drive_id", "state"],
+                metric_class=GaugeMetricFamily,
+            ),
+            Specification(
+                name="poweredgeraid_virtual_drive_cache_policy",
+                documentation="Indicates the cache policy of virtual drive",
+                labels=["controller_id", "device_group", "virtual_drive_id", "cache_policy"],
+                metric_class=GaugeMetricFamily,
+            ),
+        ]
+
+    def fetch(self) -> List[Payload]:
+        """Load the MegaRAID related information."""
+        payloads = []
+        success = self.perccli.success()
+        payloads.append(
+            Payload(
+                name="perccli_command_success",
+                labels=[],
+                value=1.0 if success else 0.0,
+            ),
+        )
+        if not success:
+            return payloads
+
+        seccesses = self.perccli.ctrl_successes()
+
+        for ctrl_id, success in seccesses.items():
+            payloads.append(
+                Payload(
+                    name="perccli_command_ctrl_success",
+                    labels=[str(ctrl_id)],
+                    value=1.0 if success else 0.0,
+                ),
+            )
+
+        controller_payload = self.perccli.get_controllers()
+        virtual_device_payload = self.perccli.get_virtual_drives()
+        payloads.append(
+            Payload(
+                name="poweredgeraid_controllers",
+                labels=[controller_payload["hostname"]],
+                value=controller_payload["count"],
+            )
+        )
+        for ctrl_id, vd_payloads in virtual_device_payload.items():
+            payloads.append(
+                Payload(
+                    name="poweredgeraid_virtual_drive",
+                    labels=[str(ctrl_id)],
```

Comment:
nitpick: if `ctrl_id` is used in many places, maybe it's worthy to return a str of `ctrl_id` or `ctrl_id = str(ctrl_id)` at the begining of the loop

Reply:
Updated